Adds support for collections

flenter opened the pull request:
Adds in:
* collections
* confirmation dialogs
* shake animation
* updated `@fiberplane/fpx-types`

and so much more.  See it in action:

https://github.com/user-attachments/assets/9ee0eb7b-8153-45c5-bdf9-e58f903df4f6

flenter committed:
WIP for groups

flenter committed:
Amazing stuff

flenter committed:
Add to group via breadcrumb

flenter committed:
Use collections

Add initial support for storing request parameters for collection items

flenter committed:
Merge remote-tracking branch 'origin/main' into initial-groups

flenter committed:
Use multiple parameters from collection item

Now supported:
* body
* query params
* path params
* headers

Note: body isn't done correctly as I need to figure out whether to store
a richer JSON format or just the body value.

flenter committed:
useRequestorStore -> useStudioStore

And also add functionality regarding setting the body (which now
takes the Content-Type header value into consideration)

flenter committed:
Minor refactor and add support for content-type

As a hint for the collection items. So the body is rendered with the
right syntax highlighting

flenter committed:
Rename app-route to item under collection in API

Also:
* streamline the create collection flow a tiny bit
* add support for deleting ~app_routes~ items from collection (to api &
  UI)
* add support for deleting a collection (in the UI)

flenter committed:
Merge remote-tracking branch 'origin/main' into initial-groups

flenter committed:
Move KeyValueParameter schema/type

And also:
Prevent 1password autocomplete create collection form

flenter committed:
Merge remote-tracking branch 'origin/main' into initial-groups

flenter committed:
wip

flenter committed:
Add redirect after adding route to collection

And some refactoring

flenter committed:
Merge remote-tracking branch 'origin/main' into initial-groups

flenter committed:
Fix build & add reset on home breadcrumb

flenter committed:
Remove stray console.log from api/inference

flenter committed:
Add proper form encoded support

Also: minor refactor to fix some hot reloading complaints

flenter committed:
Synchrnoize changes with store

flenter committed:
Sync with api

flenter committed:
Merge remote-tracking branch 'origin/main' into initial-groups

flenter committed:
Enable editing of colleciton (name)

flenter committed:
Split up CollectionSection component

as well as correct import paths for the icon component

flenter committed:
Add confirmation dialog

flenter committed:
Simplify delete confirmation

flenter committed:
Merge remote-tracking branch 'origin/main' into initial-groups

flenter committed:
Simplify PR

flenter committed:
Merge remote-tracking branch 'origin/main' into initial-groups

flenter committed:
Minor improvement/tweaks

Updated useShake to have a stable shake function (does not change
between renders so it can be used in a useEffect)

Updated empty state for collections panel
Updated not found screen for collection page
Removed some console.logs
and more...

flenter committed:
Rename ExtraRequestParams(Schema)

And remove unused props

flenter committed:
Merge remote-tracking branch 'origin/main' into initial-groups

flenter committed:
Add ai support for collection items

flenter committed:
Add position to the collection item model

flenter committed:
Add drag/drop support

flenter committed:
Add loading indicator to breadcrumb

flenter committed:
Refactor: move useStickyLoading&CollectionItemList

flenter committed:
collectionId now as a number

instead of a string

flenter committed:
Make itemId consistantly a number

flenter committed:
Split up breadcrumb components

flenter committed:
Cleanup & fix useActive...Id hooks

flenter committed:
Improve AddRouteForm

flenter committed:
Remove commented out code

flenter committed:
Use default export for tailwind config

flenter committed:
Use import instead of require in tailwind.config

flenter committed:
Merge remote-tracking branch 'origin/main' into initial-groups

brettimus commented on api/drizzle/0022_confused_maelstrom.sql:
```diff
@@ -0,0 +1,21 @@
+CREATE TABLE `collection_items` (
```
huh - so this migration is repeated but it didn't cause any issues for me locally? 🤔

flenter replied:
ah right, i don't think this isn't used,. I will remove  the other one (19)

brettimus commented on studio/tsconfig.json:
```diff
@@ -28,6 +28,6 @@
     }
   },
-  "include": ["src", "../client-library-otel/src/utils/wrapper.ts"],
+  "include": ["src", "../client-library-otel/src/utils/wrapper.ts", "src/pages/RequestorPage/CollectionForm.tsx"],
```
can revert this? 
```suggestion
  "include": ["src", "../client-library-otel/src/utils/wrapper.ts"],
```

brettimus commented on studio/tailwind.config.js:
```diff
@@ -71,12 +73,27 @@ module.exports = {
           from: { height: "var(--radix-accordion-content-height)" },
           to: { height: "0" },
         },
+        shake: {
```
the shake is back!

brettimus commented on studio/src/queries/collections.ts:
```diff
@@ -0,0 +1,316 @@
+    return error;
+  }
+
+  if (
```
could use `objectWithKey` for part of this check

flenter replied:
oh nice one.

brettimus commented on studio/src/queries/collections.ts:
```diff
@@ -0,0 +1,316 @@
+    }),
+  }).then(async (r) => {
+    if (!r.ok) {
+      console.log("r", Array.from(r.headers.entries()));
```
```suggestion
```

(remove log)

brettimus commented on studio/src/queries/collections.ts:
```diff
@@ -0,0 +1,316 @@
+          throw resultToError(result);
+        }
+
+        await r.text();
```
any reason to parse the body here? seems like we could just return

brettimus commented on studio/src/queries/collections.ts:
```diff
@@ -0,0 +1,316 @@
+
+        await r.text();
+        return undefined;
+        // return CollectionSchema.parse(json);
```
```suggestion
```

brettimus commented on studio/src/queries/app-routes.ts:
```diff
@@ -1,4 +1,4 @@
-import { useRequestorStore } from "@/pages/RequestorPage/store";
+import { useStudioStore } from "@/pages/RequestorPage/store";
```
good rename

brettimus commented on studio/src/pages/RequestorPage/store/utils.ts:
```diff
@@ -6,6 +6,7 @@ import type { RequestorState } from "./types";
 export const _getActiveRoute = (state: RequestorState): ProbedRoute => {
   return (
     state.activeRoute ?? {
+      id: -1,
```
or: `Number.NEGATIVE_INFINITY` 😄

brettimus commented on studio/src/pages/RequestorPage/store/slices/types.ts:
```diff
@@ -39,29 +41,39 @@ export interface RequestResponseSlice {
 
 export interface RoutesSlice {
-  routes: ProbedRoute[];
+  appRoutes: ProbedRoute[];
```
good rename

brettimus commented on studio/src/pages/RequestorPage/store/slices/index.ts:
```diff
@@ -2,6 +2,7 @@ export { aiSlice } from "./aiSlice";
 export { routesSlice } from "./routesSlice";
 export { tabsSlice } from "./tabsSlice";
-export type { Store } from "./types";
+export type { StudioState as StudioStore } from "./types";
```
StudioState or StudioStore?

flenter replied:
Renaming in vscode has its limitations :) ... I noticed that in slices i also exported it differently. The idea is to call it call it state  and we keep it in a store

brettimus commented on api/src/db/schema.ts:
```diff
@@ -175,3 +175,61 @@ export const aiRequestLogs = sqliteTable("ai_request_logs", {
+  id: integer("id", { mode: "number" }).primaryKey({ autoIncrement: true }),
+  name: text("name").notNull().unique(),
+  createdAt: text("created_at").notNull().default(sql`(CURRENT_TIMESTAMP)`),
+  updatedAt: text("updated_at").notNull().default(sql`(CURRENT_TIMESTAMP)`),
```
would suggest using the "timetamps" helper since it'll be iso8601

```suggestion
  ...timestamps,
```

brettimus commented on studio/src/App.tsx:
```diff
@@ -23,22 +18,15 @@ export function App() {
+                element={<AiRequestLogsPage />}
               />
               <Route
-                path={REQUEST_DETAILS_TRACE_ROUTE}
```
bye bye request details page

brettimus commented on studio/src/Layout/Settings/SettingsScreen.tsx:
```diff
@@ -1,6 +1,6 @@
 import { Button } from "@/components/ui/button";
 import { SettingsPage } from "@/pages/SettingsPage";
-import { Icon } from "@iconify/react/dist/iconify.js";
+import { Icon } from "@iconify/react";
```
was there a reason we were importing from `dist/iconify.js`?

flenter replied:
Not sure, it's vscode that imports it like that, but to me this seems brittle

brettimus commented on studio/src/hooks/useStickyLoading.ts:
```diff
@@ -0,0 +1,21 @@
+    return () => clearTimeout(timeout);
+  }, [loading, duration]);
+
+  return sticky;
```
should this return `sticky || loading`? 

if i understand correctly, it is a minimum duration for loading, but if `loading` is still true after `duration`, then it should still return `true`

brettimus reviewed (commented):
first round of comments -

brettimus commented on packages/types/src/api.ts:
```diff
@@ -11,3 +11,40 @@ export const TraceListResponseSchema = z.array(TraceSummarySchema);
 
 export type TraceDetailSpansResponse = Array<OtelSpan>;
+
+export const CollectionSchema = z.object({
```
Important! Gotta bump the version in `packages/types/package.json`

then edit the pull request description to remind us to publish the new version of fpx-types

brettimus commented:
Reminders:

- Update changelog
- Update api/package.json version
- Update packages/types/package.json version

Must remember to publish packages/types before releasing our next canary of studio

flenter committed:
Initial feedback on PR

flenter committed:
Fix typo in path

flenter committed:
And now it's all formatted nicely

flenter committed:
Bump versions

flenter committed:
Avoid false loading state on long request

flenter committed:
Format package.json

flenter committed:
Use strings as item index for dnd

There was an issue with dnd-kit when using index as the (actual) number
type.

The dragging animation wouldn't work

brettimus commented on studio/src/hooks/useInputFocusDetection.ts:
```diff
@@ -19,11 +16,12 @@ export function useInputFocusDetection() {
-    document.addEventListener("focus", handleFocus, { capture: true, signal });
-    document.addEventListener("blur", handleBlur, { capture: true, signal });
+    document.addEventListener("focus", handleFocus, { capture: true });
+    document.addEventListener("focus", handleFocus, { capture: true });
```
looks like the focus handler is registered twice here instead of also registering the `blur` handler

brettimus commented:
~bug: clicking on the "filter routes" text input does not activate focus on the component for me anymore ((haven't tried to figure out the cause yet))~

<img width="350" alt="image" src="https://github.com/user-attachments/assets/3d33b72b-060e-4701-86a1-08f079da5c09" />


~keyboard shortcut still works for me~

***

fixed, needed to pass through click events from the placeholder elements

brettimus commented on studio/src/pages/RequestorPage/AddRouteToCollection/AddToRouteForm.tsx:
```diff
@@ -0,0 +1,49 @@
+  const { data: collections, isLoading } = useCollections();
+  const { activeRoute } = useStudioStore("activeRoute");
+
+  // No active route? Then there's nothing to do here
```
could be good to at least put a message in the UI here — otherwise if the user has typed something that doesn't match the routes, then all they see is an empty popover 

<img width="742" alt="image" src="https://github.com/user-attachments/assets/3dfddeb2-4729-4a25-a685-ea5ad59cbe0b" />

brettimus commented on studio/src/pages/RequestorPage/AddRouteToCollection/AddToRouteForm.tsx:
```diff
@@ -0,0 +1,49 @@
+  }
+
+  if (collections.length === 0) {
+    return <div>Empty</div>;
```
probably for future, but maybe we always initialize a "starter" collection of some sort? 

maybe for now could just put some helper text like "Create a collection to get started" or something

brettimus commented on studio/src/pages/RequestorPage/RequestPanel/RequestPanel.tsx:
```diff
@@ -223,13 +251,18 @@ export const RequestPanel = memo(function RequestPanel(
           keyPlaceholder="header-name"
           keyValueParameters={requestHeaders}
           onChange={(headers) => {
-            setRequestHeaders(headers);
```
👀

brettimus replied:
i think we need this line

brettimus committed:
Remove noisy log from prev pr

brettimus committed:
Remember to set request headers

brettimus committed:
Remove dead code

brettimus commented on studio/src/utils/index.ts:
```diff
@@ -216,44 +224,79 @@ export function isSensitiveEnvVar(key: string) {
+  });
+}
+
+export function createObjectFromKeyValueParameters<
```
as far as i can tell, this is only used in cases where we sync with a collection - however, one thing to be wary of is that this function ignores whether a parameter is enabled/disabled. which is to say, it will copy/sync parameters that have been disabled by the end user

brettimus committed:
Factor out a common piece of code for error handling in queries/collections

brettimus commented on studio/src/queries/collections.ts:
```diff
@@ -0,0 +1,287 @@
+  return mutation;
+}
+
+async function responseToJsonOrText(r: Response) {
```
added this to factor out the ternary that was happening in most `if (!res.ok) { ... }` blocks

brettimus commented on studio/src/pages/RequestorPage/store/utils.ts:
```diff
@@ -6,6 +6,7 @@ import type { RequestorState } from "./types";
 export const _getActiveRoute = (state: RequestorState): ProbedRoute => {
   return (
     state.activeRoute ?? {
+      id: Number.NEGATIVE_INFINITY,
```
hahaha i was being cheeky when i suggested `NEGATIVE_INFINITY` but i like that it's here

brettimus committed:
Properly register blur handler and fix Search input

brettimus committed:
Add message if no matching route

brettimus committed:
Remove dead code and tweak copy on collection add form

brettimus committed:
Tidy collections empty state

brettimus committed:
tidy required message

brettimus committed:
Tweak collection empty state and loading state copy

brettimus committed:
Tighten up spacing on breadcrumbs

brettimus committed:
Fix overflow issues on AddRouteForm.tsx

brettimus committed:
Only show registered routes in add routes menu

brettimus committed:
Fix text truncation in the collections panel

brettimus committed:
Only create objects from enabled key value params

brettimus committed:
Move around custom routes

brettimus committed:
Use zValidator for the POST /v0/collections route

brettimus committed:
Change error message styles

brettimus committed:
Add an app.onError handler and then simplify the routes/collections error handling

brettimus committed:
Remove unnecessary error handling

brettimus committed:
So much more shadow for that drag

brettimus committed:
Ignore new script

brettimus committed:
Merge branch 'main' into initial-groups

brettimus committed:
Format mcp-server package.json

brettimus commented on studio/src/pages/RequestorPage/CollectionSection/CollectionItemListItem.tsx:
```diff
@@ -0,0 +1,160 @@
+                collectionId,
+                itemId,
+                extraParams: {
+                  position: Number.MAX_SAFE_INTEGER,
```
this causes problems 😅  - it will always be at the bottom, even when i try to reorder something below it

brettimus replied:
okay, issue was on the api side! fixed

brettimus committed:
Fix issue where "move to bottom" logic would stick the item to the bottom

brettimus committed:
Start investigating race condition that flashes UI to different states

brettimus commented on studio/src/pages/RequestorPage/RequestorPageContent/RequestorPageContent.tsx:
```diff
@@ -32,12 +42,93 @@ type RequestorPageContentProps = {
+
+    setRequestParams(item.app_requests);
+  }, [item, setRequestParams]);
+
```
something is a little wonky with the introduction of these effects — i'm getting some flash-in behavior, especially where the UI will temporarily show the "activeResponse", then flash to the trace response. 

aside from that, the headers will temporarily flash to the list of headers made in the proxied request, then flash back to the og headers in the form. 

will try to resolve today

brettimus replied:
at any rate, the effects could probably use comments that explain their purpose

brettimus replied:
removing the second effect in this highlighted code did the trick for me. i'm not sure if it served a purpose, trying to look for things i might have broken

brettimus committed:
Fix focus ring clipping

brettimus committed:
Run format and remove rogue logs

brettimus committed:
Maintain consistent height in the MainTopSection

brettimus committed:
Trigger the collection popover with keyboard shortcut

brettimus committed:
Add tooltip for collection button

brettimus committed:
Bump studio to a canary version real quick

brettimus committed:
Update db schema

brettimus committed:
Remove oddly dead code

brettimus committed:
Temp mod to gitignore

brettimus committed:
Fix issue where x-fpx-trace-id was sneaking into the headers for historical requests in the ui
